### ISSUE TITLE
Read-Only Observable

### ISSUE THREAD

**Issue by AnthonyMDev** [SwiftBond/Bond#198]
How do I implement a read-only observable property. Because `value` on `Observable`  is read-write, I can't seem to find a way to stop another class from changing my observable's `value`, even if the observable property is read-only itself.

**Comment by srdanrasic** [comment]
Yeah, that's because Observable is a class. You could do this by introducing a struct wrapper around Observable.

```swift
public struct MutableObservable<Wrapped>: EventProducerType {
  
  private var observable: Observable<Wrapped>
  
  public var value: Wrapped {
    get {
      return observable.value
    }
    set {
      observable.value = newValue
    }
  }
  
  public init(_ value: Wrapped) {
    observable = Observable(value)
  }
  
  public var replayLength: Int {
    return observable.replayLength
  }

  public func observe(observer: Wrapped -> Void) -> DisposableType {
    return observable.observe(observer)
  }
}
```

Than, you would use it like this:

```swift
public private(set) var name = MutableObservable("Jim")
```

**Comment by manas-chaudhari** [comment]
I am facing a similar issue with `EventProducer`. Another struct/class can be created for EventProducer.

I feel that outputs of rx operators like `map`, `filter`, `combine` should be immutable event producers as they are like derived properties. Only the source where an `EventProducer`/`Observable` is constructed should have access to `next`/`value`. 
@srdanrasic: What do you think?

**Comment by AnthonyMDev** [comment]
@srdanrasic Thanks, I understood that it was due to `Observable` being a class. Making a struct wrapper around the class is a great idea. I feel that this would be a good addition to the core library. Is there any interest in including this, or should I just go ahead and make the struct in my own project?

**Comment by srdanrasic** [comment]
Yeah, that would be a good idea :)

**Comment by Danappelxx** [comment]
:+1: I'm all for it

**Comment by ivanmoskalev** [comment]
@AnthonyMDev awesome, let's do it!

**Comment by AnthonyMDev** [comment]
Great. @srdanrasic Do you want to make a PR since you basically already wrote something in the above post? I haven't actually implemented this in my project yet, but I'd be happy to work on it if you don't!

**Comment by srdanrasic** [comment]
Let me do it.

**Comment by manas-chaudhari** [comment]
@srdanrasic I am failing to understand how this is solving the problem. Please explain.
Here's what I tried:
```
class ViewModel {
    private(set) var readOnlyVariable: MutableObservable<String> = MutableObservable("inner")

    init() {
        readOnlyVariable.observe(on: ImmediateExecutionContext) { (e) -> () in
            print(e) // Prints 'value from outside'
        }
    }
}

ViewModel().readOnlyVariable.value = "value from outside"
```

I am able to change the value of the observable from outside the class.
If I use `let`, then it cannot be modified. But then, it cannot be modified from inside also.

**Comment by srdanrasic** [comment]
@manas-chaudhari, `private` in Swift refers to file. When you set something to private it is visible only within that file. Playground is just one file. Had your ViewModel been defined in a separate file, this would work.

**Comment by manas-chaudhari** [comment]
I tried with separate files and it works.
Thanks.
